### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] misleading error messages

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9664]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=GYVwdgxgLglg9mABABwIYCcDOBTAUphAHgBVFsAPKbMAE00UynRjAHMA%2BACheRCgC4GTFqwCUg0gG8AUIkTpsUEOiS4AygHkAcgDo0WbNzC8ooxKnrEA3NIC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
// T is used twice - maybe it should say "once in the signature"?
function parseJson<T extends string>(input: string): T {
  return JSON.parse(input) as T;
}

// T is used zero times
function parseJson<T extends string>(input: string): any {
  return JSON.parse(input) as any;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect the error message should reflect correctly the number of times the type parameter is used.

### Actual Result

"Type parameter T is used only once." in both cases

### Additional Info

Let's also include a doc change to the docs page, if counting parameters is only supposed to be counting the usages in the signature?

**Comment by That-Guy977** [comment]
The current behavior does analyze the body to some extent.
```diff
  function bar<T>(_: T) {
    const x: T = /* ... */
+   const y: T = /* ... */
  }
```
Without the highlighted line, `Type parameter T is used only once.` is reported, but adding it suppresses the error. [Playground](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcA8AVAfACgPoBciGAlIgN4BQiiECAzlIgB5EaIC8iYIANrwEIA3NVoMmATzadufQSIC%2BlSqEiwEiAEYBDAE6ZchYmSo06YRi2lce-YZSVA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by JoshuaKGoldberg** [comment]
> `+   const y: T = /* ... */`

That ... might be a bug in the rule. I don't think it's intentional that anything in the function _body_ that's not a part of the _signature_ should change how the rule interprets the _signature_. Filed #9735.
